Paco-Server/src/com/google/sampling/experiential/server/EventRetriever.java

```diff
@@ -132,29 +151,60 @@ public void postEvent(String who, String lat, String lon, Date whenDate, String
     Event event = new Event(who, lat, lon, whenDate, appId, pacoVersion, what, shared,
         experimentId, experimentName, experimentVersion, responseTime, scheduledTime, blobs, tz,
         groupName, actionTriggerId, actionTriggerSpecId, actionId);
-    Transaction tx = null;
-    try {
-      tx = pm.currentTransaction();
-      tx.begin();
-      pm.makePersistent(event);
-      if (isJoinEvent) {
-        ExperimentAccessManager.addJoinedExperimentFor(who, Long.valueOf(experimentId), responseTime);
-      } else if (!isScheduleEvent && !isStopEvent) {
-        new ParticipationStatsService().updateResponseCountWithEvent(event);
+    //persistInCloudSql flag will determine which flow to go. Flow 1:persist in data store and send event to the cloud sql queue 
+    // Flow 2: persist in cloud sql
+    if(persistInCloudSql && eventJson!=null){
+      Long id;
+      try {
+        id = eventJson.getLong("id");
+        event.setId(id);
+      } catch (JSONException e) {
+        log.warning("Event json id parsing error"+e);
+      }
+      try{
+        cloudSqlDaoImpl.insertEvent(event);
+      } catch (SQLException sqle){
+        log.warning("Exception while inserting data into cloudsql db"+sqle);
       }
-      tx.commit();
-      log.info("Event saved");
-    } finally {
-      if (tx.isActive()) {
-        log.info("Event rolled back");
-        tx.rollback();
+
+    }else{
+      
+      Transaction tx = null;
+      try {
+        tx = pm.currentTransaction();
+        tx.begin();
+        pm.makePersistent(event);
+        event.setWhat(what);
+        if (isJoinEvent) {
+          ExperimentAccessManager.addJoinedExperimentFor(who, Long.valueOf(experimentId), responseTime);
+        } else if (!isScheduleEvent && !isStopEvent) {
+          new ParticipationStatsService().updateResponseCountWithEvent(event);
+        }
+        tx.commit();
+        log.info("Event saved in datastore");
+       
+      } finally {
+        if (tx.isActive()) {
+          log.info("Event rolled back");
+          tx.rollback();
+        }
+        pm.close();
       }
-      pm.close();
+      //TODO confirm if this should be outside of transaction
+      sendToCloudSqlQueue(eventJson, event.getId());
```

Comment:
From the google.cloud.com page, what I infer is that the application that creates the task, is not notified on its status, whether it completed or not. So, I don't think it should be inside the transaction.